File: Go/day2_part2.go
package main 

import (
	"fmt"
	"encoding/csv"
	"os"
	"errors"
)

func day2Part2() {
	reader := csv.NewReader(func() *os.File {
		f, _ := os.Open("../day2_input.txt")
		return f
	}())

	reader.Comma = '	'

	data, _ := reader.ReadAll()
	total := 0
	for _, row := range data {
		intRow := toInt(row)
		for i, num := range intRow {
			j, err := findDivisible(num, intRow[i+1:])
			if err == nil {
				total += j
			}
		}
	}
	fmt.Println(total)
}

func findDivisible(i int, s []int) (int, error) {
	for _, num := range s {
		if i%num == 0 {
			return i/num, nil
		}
		if num%i == 0 {
			return num/i, nil
		}
	}
	return 0, errors.New("none found")
}

File: Kotlin/day2_part2.kt
import com.opencsv.CSVParserBuilder
import com.opencsv.CSVReaderBuilder
import java.io.*
import java.util.*

fun main(args: Array<String>) {
    val parser = CSVParserBuilder().withSeparator('\t').withIgnoreQuotations(true).build()
    val reader = CSVReaderBuilder(FileReader("../day2_input.txt")).withCSVParser(parser).build()
    var total = 0
    while(true) {
        val row = reader.readNext()?.map { it.toInt() }?.toTypedArray() ?: break
        row.forEachIndexed { i, num ->
            val slice = row.sliceArray(IntRange(i+1, row.size-1))
            val res = findDivisible(num, slice)
            total += if(res != null) res else 0
        }
    }
    println(total)
}

fun findDivisible(i: Int, s: Array<Int>): Int? {
    s.forEach{
        if(i%it == 0){
            return i/it
        }
        if(it%i == 0){
            return it/i
        }
    }
    return null
}